### ISSUE TITLE
Raising StopIteration inside afunc that is passed to lift_to_async() leaves the Future pending forever

### ISSUE THREAD

**Issue by alex-ber** [alex-ber/AlexBerUtils#14]
From run_in_executor() of langchain-core runnable/config.py + get_executor_for_config()

```python



@contextmanager
def get_executor_for_config(
    config: Optional[RunnableConfig],
) -> Generator[Executor, None, None]:
    """Get an executor for a config.

    Args:
        config (RunnableConfig): The config.

    Yields:
        Generator[Executor, None, None]: The executor.
    """
    config = config or {}
    with ContextThreadPoolExecutor(
        max_workers=config.get("max_concurrency")
    ) as executor:
        yield executor


async def run_in_executor(
    executor_or_config: Optional[Union[Executor, RunnableConfig]],
    func: Callable[P, T],
    *args: P.args,
    **kwargs: P.kwargs,
) -> T:
    """Run a function in an executor.

    Args:
        executor_or_config: The executor or config to run in.
        func (Callable[P, Output]): The function.
        *args (Any): The positional arguments to the function.
        **kwargs (Any): The keyword arguments to the function.

    Returns:
        Output: The output of the function.

    Raises:
        RuntimeError: If the function raises a StopIteration.
    """

    def wrapper() -> T:
        try:
            return func(*args, **kwargs)
        except StopIteration as exc:
            # StopIteration can't be set on an asyncio.Future
            # it raises a TypeError and leaves the Future pending forever
            # so we need to convert it to a RuntimeError
            raise RuntimeError from exc

    if executor_or_config is None or isinstance(executor_or_config, dict):
        # Use default executor with context copied from current context
        return await asyncio.get_running_loop().run_in_executor(
            None,
            cast(Callable[..., T], partial(copy_context().run, wrapper)),
        )

    return await asyncio.get_running_loop().run_in_executor(executor_or_config, wrapper)

```

The wrapper part should be integrated into lift_to_async().

**Comment by alex-ber** [comment]
FIx:

```python

def lift_to_async(afunc, /, *args, **kwargs):
    """
    Executes an asynchronous function in a synchronous context preserving ContextVar's context.

    Args:
        afunc: The asynchronous function to be executed.
        *args: Positional arguments to pass to the function.
        **kwargs: Keyword arguments to pass to the function.

    Returns:
        The result of the asynchronous function call.

    Raises:
        RuntimeError: If the function raises a StopAsyncIteration.
    """
    logger.info("lift_to_async()")

    def wrapper():
        try:
            return afunc(*args, **kwargs)
        except StopAsyncIteration as exc:
            raise RuntimeError("Async generator exhausted unexpectedly") from exc

    ctx = contextvars.copy_context()
    afunc_call = functools.partial(ctx.run, wrapper)
    result = _execute_async_in_sync(afunc_call)
    return result

```

See https://chatgpt.com/share/6733c942-62b8-8007-87d9-d668e667e59e

**Comment by alex-ber** [comment]
Fixed in 0.12.0.
